### ISSUE TITLE
Add NFD, NFC, NFKD and NFKC normalization

### ISSUE THREAD

**Issue by bungle** [simdutf/simdutf#553]
It would be great to have normalization function(s):
See: https://unicode.org/reports/tr15/

This is somewhat duplicate to #80, but a more specific one.

**Comment by the-moisrex** [comment]
Since I've [implemented Normalization recently](https://github.com/the-moisrex/webpp/blob/c3528094734c21cbb3ee95378a27f677058162d9/webpp/unicode/normalization.hpp) and I saw this issue, I have to say that SIMDifying Unicode Normalization algorithms is not a trivial task; and I haven't seen it done before.

NFD, and NFC alone (for my implementation) requires 55 KiB (10 for CCC tables, 31 for Decompositions, and 14 for Compositions) of lookup table gibberish.

Hangul code points are handled algorithmically, but the rest is done by table lookups.

They also require re-ordering which swaps code points, which again, it's not a SIMD-friendly algorithm; but hey, I'm not really good at SIMD anyway, so who knows!

The [IDNA Library](https://github.com/ada-url/idna) from ada (which @lemire himself is one of the authors) includes NFC and NFD also.

**Comment by clausecker** [comment]
I've considered this too.  The main way SIMD would seem to help here is to have a fast path for when no normalisation is needed.  If the SIMD kernel detects something that needs normalisation, it would bail out to some scalar code.  I suppose most text does not require any modifications to be normalised, so I guess it could be worthwhile.

**Comment by the-moisrex** [comment]
I'm going to assume there's not that many windows of chances where we can bet the fast path on. Here's [some stats](https://github.com/the-moisrex/codeshells/blob/b8117a1202d45f219205b7a923316d3628550ead/awk.idna-stats/idna-calc.awk) on `IDNAMappingsTable.txt`:

> Total code points (up to U+10ffff): 1114112
> Mapped code points: 6348 (0.57%)
> Valid code points: 148191 (13.30%)
> Disallowed code points: 959275 (86.10%)
> Ignored code points: 294 (0.03%)
> Other code points: 4 (0.00%)
> Covered code points in file: 1114112
> Not valid code points: 965921 (86.70%)
> Uncovered code points: 0

> Total code points (up to U+3347f): 210048
> Mapped code points: 6348 (3.02%)
> Valid code points: 148191 (70.55%)
> Disallowed code points: 55451 (26.40%)
> Ignored code points: 54 (0.03%)
> Other code points: 4 (0.00%)
> Covered code points in file: 210048
> Not valid code points: 61857 (29.45%)
> Uncovered code points: 0


Yep, 86% of code points are non-valid. Your best chance of avoiding to go through the normalization is to check for ASCII.

I get your fast-path idea, but Unicode is too effing random.

**Comment by the-moisrex** [comment]
> These percentages don't really matter.  What matters is how often the disallowed code points appear in real text.

Yes.


BTW, the stats are for IDNA, not decomposition. Sorry about that.

**Comment by h-2** [comment]
Yeah, I also think that most of our inputs and definitely most windows of most inputs are valid. Just being able to check that efficiently and then performing normalisation on the text windows that need fixing would already be great.

On a related note: utf8_to_lower would likely be in a similar situation, right? Or would that be easier because the number of codepoints (and the bytes per codepoint) would stay the same?¹

¹Is that assumption even correct?

**Comment by clausecker** [comment]
> On a related note: utf8_to_lower would likely be in a similar situation, right? Or would that be easier because the number of codepoints (and the bytes per codepoint) would stay the same?¹

`to_lower` is probably harder as characters have the habit of being in upper case every once in a while.  The number of bytes per code point does not necessarily stay the same, e.g. Turkish upper case `İ` (U+0130, two bytes) becomes `i` (U+0069, one byte) when translated to lower case.  Another example would be German `ẞ` (U+1E9E, three bytes) becoming `ß` (U+00DF, two bytes) when translated to lower case.

**Comment by WojciechMula** [comment]
> On a related note: utf8_to_lower would likely be in a similar situation, right? Or would that be easier because the number of codepoints (and the bytes per codepoint) would stay the same?¹
> 
> ¹Is that assumption even correct?

Unfortunately, changing case of letters may change the number of UTF-8 bytes *as well* as the number of codepoints (for instance uppercase of ligature 'ffl' is a three-character string of "FFL").

There are relatively small number of characters having uppercase/lowercase counterparts, and among them only a little fraction have this unfriendly feature of different lengths.

**Comment by h-2** [comment]
Thanks for the examples 💡 

Yeah, seems like it would be non-trivial. Seems like it will be more and more difficult for me to avoid going from UTF8 to UTF32 and then going codepoint by codepoint and then back 😿

**Comment by WojciechMula** [comment]
> Thanks for the examples 💡
> 
> Yeah, seems like it would be non-trivial. Seems like it will be more and more difficult for me to avoid going from UTF8 to UTF32 and then going codepoint by codepoint and then back 😿

I think it would be very difficult to stay in UTF-8 when you need case changing. I mean, these mappings are irregular and you cannot express them nicely as some vectorized expressions. Not to mention that new Unicode versions may add new characters to the mapping. But what you can for sure do, is to handle ASCII chunks separately, as case changing is trivial in vectorized code.

You may find this a bit useful: http://0x80.pl/notesen/2025-02-02-utf32-change-case.html
